source/DamageProfile.cpp

```diff
@@ -135,9 +135,51 @@ void DamageProfile::PopulateDamage(DamageDealt &damage, const Ship &ship) const
 	// Hull damage is blocked 100%.
 	// Shield damage is blocked 0%.
 	damage.shieldDamage *= shieldFraction;
-	damage.hullDamage = (weapon.HullDamage()
-		+ weapon.RelativeHullDamage() * attributes.Get("hull"))
-		* ScaleType(1., attributes.Get("hull protection"));
+	double hull = ship.HullLevel();
+	double minHull = ship.MinimumHull();
+	if(hull < minHull)
+	{
+		double healPart = minHull - hull - 0.25d;
+		damage.hullDamage = (weapon.DisabledDamage()
+			+ weapon.RelativeDisabledDamage() * attributes.Get("hull"))
+			* ScaleType(1., attributes.Get("hull protection"));
+
+		// Test if the weapon is healing and if it will reach the disabled threshold.
+		if(-damage.hullDamage > healPart)
+		{
+			double hullFraction = (damage.hullDamage + healPart) / damage.hullDamage;
+
+			double hullDamage = (weapon.HullDamage()
+			+ weapon.RelativeHullDamage() * attributes.Get("hull"))
+			* ScaleType(1., attributes.Get("hull protection")) * hullFraction;
+
+			// hullDamage only applies here if it heals.
+			if(hullDamage < 0)
+				damage.hullDamage += hullDamage;
+		}
+	}
+	else
+	{
+		// The hull is above the disabled hull threshold.
+		double nonDisabledHull = hull - minHull + .25d;
+		damage.hullDamage = (weapon.HullDamage()
+			+ weapon.RelativeHullDamage() * attributes.Get("hull"))
+			* ScaleType(1., attributes.Get("hull protection"));
+
+		// Test if damage will pass disabled threshold.
+		if(damage.hullDamage > nonDisabledHull)
+		{
+			double disabledFraction = nonDisabledHull / damage.hullDamage;
+
+			double disabledDamage = (weapon.DisabledDamage()
+			+ weapon.RelativeDisabledDamage() * attributes.Get("hull"))
+			* ScaleType(1., attributes.Get("hull protection")) * disabledFraction;
+
+			// disabledDamage only applies here if it damages.
+			if(disabledDamage >= 0)
+				damage.hullDamage += disabledDamage;
+		}
+	}
```

Comment:
I think this is incorrect; you're adding disabledDamage to hullDamage after setting hullDamage to the normal amount, when we should be reducing hullDamage in some way.
```suggestion
	damage.hullDamage = (weapon.HullDamage()
		+ weapon.RelativeHullDamage() * attributes.Get("hull"))
		* ScaleType(1., attributes.Get("hull protection"));
	double hull = ship.HullLevel() - ship.MinimumHull();
	if(hull > 0.)
	{
		if(damage.hullDamage > hull)
		{
			damage.hullDamage *= hull / damage.hullDamage;
			damage.hullDamage += (weapon.DisabledDamage()
				+ weapon.RelativeDisabledDamage() * attributes.Get("hull"))
				* ScaleType(1., attributes.Get("hull protection"));
		}
	}
```
The above code should handle hull damage as shield damage is handled. We check if the hull damage would push us past the disabled threshold (as we check if shield damage would deplete the shields), then multiply the hull damage by the hull fraction (as we multiply shield damage by the shield fraction). *Then* we add the disabled damage to the hull damage.

This doesn't account for "healing" weapons, but neither does shield damage, so I don't see the point in accounting for that here as was done previously.

Reply:
If we changed HullLevel to return the hull level remaining until disabled (as Hull returns the hull fraction until disabled) then we wouldn't need to make MinimumHull public and the `double hull = ship.HullLevel() - ship.MinimumHull();` line could be changed to just `double hull = ship.HullLevel();`.

Reply:
> This doesn't account for "healing" disabled damage weapons, but healing weapons aren't otherwise handled differently, so I don't see the point in accounting for that here as was done previously.

Well, the "healing" disabled damage weapons were discussed in #4394 as a nice feature to add, but this PR is already valuable when they are not in yet, so I'm okay to keep that for a follow-up PR.

Reply:
> DisabledHullLevel()

I guess it should be called HullUntilDisabled(), but I agree with the rest of your comment.